server/sonar-server/src/main/java/org/sonar/server/computation/measure/MeasureImpl.java

```diff
@@ -36,40 +42,100 @@
   private final Level dataLevel;
   @Nullable
   private QualityGateStatus qualityGateStatus;
+  @Nullable
+  private MeasureVariations variations;
+  @Nullable
+  private String description;
 
-  protected MeasureImpl(ValueType valueType, @Nullable Double value, @Nullable String data, @Nullable Level dataLevel) {
+  private MeasureImpl(ValueType valueType, @Nullable Integer ruleId, @Nullable Integer characteristicId, @Nullable Double value, @Nullable String data, @Nullable Level dataLevel) {
     this.valueType = valueType;
+    this.ruleId = ruleId;
+    this.characteristicId = characteristicId;
     this.value = value;
     this.data = data;
     this.dataLevel = dataLevel;
   }
 
-  public static MeasureImpl create(boolean value, @Nullable String data) {
-    return new MeasureImpl(ValueType.BOOLEAN, value ? 1.0d : 0.0d, data, null);
+  public static Builder builder() {
+    return new Builder();
   }
 
-  public static MeasureImpl create(int value, @Nullable String data) {
-    return new MeasureImpl(ValueType.INT, (double) value, data, null);
-  }
+  public static final class Builder {
+    public static final String RULE_AND_CHARACTERISTIC_ERROR_MSG = "A measure can not be associated to both a Characteristic and a Rule";
+    private Integer ruleId;
+    private Integer characteristicId;
 
-  public static MeasureImpl create(long value, @Nullable String data) {
-    return new MeasureImpl(ValueType.LONG, (double) value, data, null);
-  }
+    /**
+     * Sets the rule this measure is associated to.
+     *
+     * @throws UnsupportedOperationException if the characteristicId as already been set
+     *
+     * @see #forCharacteristic(int)
+     */
+    public Builder forRule(int ruleId) {
+      if (characteristicId != null) {
+        throw new UnsupportedOperationException(RULE_AND_CHARACTERISTIC_ERROR_MSG);
+      }
+      this.ruleId = ruleId;
+      return this;
+    }
 
-  public static MeasureImpl create(double value, @Nullable String data) {
-    return new MeasureImpl(ValueType.DOUBLE, value, data, null);
-  }
 
-  public static MeasureImpl create(String value) {
-    return new MeasureImpl(ValueType.STRING, null, requireNonNull(value), null);
+    /**
+     * Sets the characteristic this measure is associated to.
+     *
+     * @throws UnsupportedOperationException if the ruleId as already been set
+     *
+     * @see #forCharacteristic(int)
+     */
+    public Builder forCharacteristic(int characteristicId) {
+      if (ruleId != null) {
+        throw new UnsupportedOperationException(RULE_AND_CHARACTERISTIC_ERROR_MSG);
+      }
+      this.characteristicId = characteristicId;
+      return this;
+    }
+
+    public MeasureImpl create(boolean value, @Nullable String data) {
```

Comment:
@henryju had pointed me to such a Measure... I don't remember which one right now, but I might be able to find it